src/ruby/pb/test/xds_client.rb

```diff
@@ -95,17 +96,45 @@ def get_client_stats(req, _call)
       end
       $watchers.delete_at($watchers.index(watcher))
     end
+    # convert results into proper proto object
+    rpcs_by_method = {}
+    watcher['rpcs_by_method'].each do | rpc_name, rpcs_by_peer |
+      rpcs_by_method[rpc_name] = LoadBalancerStatsResponse::RpcsByPeer.new(
+        rpcs_by_peer: rpcs_by_peer
+      )
+    end
     LoadBalancerStatsResponse.new(
+      rpcs_by_method: rpcs_by_method,
       rpcs_by_peer: watcher['rpcs_by_peer'],
       num_failures: watcher['no_remote_peer'] + watcher['rpcs_needed']
     );
   end
 end
 
+# execute 1 RPC and return remote hostname
+def execute_rpc(op, fail_on_failed_rpcs)
+  remote_peer = ""
+  begin
+    op.execute
+    if op.metadata.key?('hostname')
```

Comment:
Where is the `hostname` metadata key getting populated?

Reply:
This is the updated spec for this xDS interop test - that the remote `hostname` will be returned as part of the call's _initial_ metadata, instead of in the RPC's response proto.

In the past, the remote `hostname` is being returned as a field in the `SimpleResponse` proto https://github.com/grpc/grpc/blob/v1.31.x/src/ruby/pb/test/xds_client.rb#L127, https://github.com/grpc/grpc/blob/master/src/proto/grpc/testing/messages.proto#L120. But since now, after the updated spec, we could be sending `EmptyCall` RPC, we can't be expecting this field to be in the response proto uniformly. So the updated spec specified that this key will be populated as part of the initial metadata.

Also, an RPC may be designed to not be successful as part of the test, so we need to check whether that key exists in the initial metadata as well.

Reply:
Got it, thanks.

> Also, an RPC may be designed to not be successful as part of the test, so we need to check whether that key exists in the initial metadata as well.


Slightly confused about this part, though. Looks like we currently don't check the peer via metadata at all if the call fails.

Reply:
The spec mentioned that `All clients read hostname from header[“hostname”] instead of from RPC response`. It seems that it's unclear whether the initial metadata should be read if the RPC failed. I will follow up with Menghan and Eric to see what the behavior should be. As it stands, it passes all the test cases so perhaps it doesn't matter for now. I am inclined to merge this PR as is. If there's an issue with this particular piece of logic we can always fix it later.

Reply:
Right, whether to read "hostname" from metadata when the RPC fails is  currently not defined. Mostly because none of the existing tests depends on hostname when the RPC fails. It's OK for the client to ignore it.

This may change in the future, to test failure cases (fault injection for example). But it not super clear how the behavior will be, so let's worry about it later.